### ISSUE TITLE
Why use JavaScriptCore and not v8 like nodejs does?

### ISSUE THREAD

**Issue by maticrivo** [voodooattack/nexusjs#6]

**Comment by voodooattack** [comment]
Hi! Thanks for your interest in the project!

V8 uses something called Isolates to lock the entire virtual machine during code execution on any thread. This essentially means that there is no way for multiple threads to concurrently call into a JavaScript context with V8.

JavaScriptCore locks specific contexts on the other hand (where a context is a *part* of the virtual machine, and consists of a stack pointer and packaged closure variables). If two contexts don’t share a closure, they can both run in parallel with no contention on different threads. 

In the case where two contexts share a global/closure variable, one will acquire the lock and the other thread will have to wait, thus introducing atomic behaviour. 

So, to summarise: V8 is simply not *capable* of supporting shared mutable state with its current approach and design philosophy. It would have to be rewritten from scratch to support such a use case.

JavaScriptCore’s design provides a granular approach when it comes to concurrent locking, and this is why we’re using it for Nexus.js.

**Comment by maticrivo** [comment]
great! thanks so much for the answer! i'm gonna play around with it and see if i can run some of my projects on it :)

**Comment by voodooattack** [comment]
You're welcome.

**Comment by dou4cc** [comment]
How about [rust-mozjs](https://github.com/servo/rust-mozjs)? It is brand new and maybe faster?

**Comment by voodooattack** [comment]
@dou4cc Why? WebKit is perfect for our use case and is already heavily ingrained into the code. Why would I want to switch engines now?

**Comment by dou4cc** [comment]
@voodooattack I was just asking your opinion on rust-mozjs, since it seems not easy to unbind the JS engine with Nexus.js.

**Comment by voodooattack** [comment]
@dou4cc I see. Well, I have no qualms about it. I just have no desire to rewrite everything from scratch just to find out if it is multi-threading compatible or not.

My first attempt with SpiderMoney (that was before I tried WebKit) was disastrous. I'm not sure if Firefox Quantum brought multi-threading support to SpiderMonkey or not, but I'm not going to test it.
